File: Core/src/OftSecurityMode.cs
namespace BlueHeighliner.OpenFrameTransport;

/// <summary>
/// How a connection (or an <see cref="IOftPeer"/>'s connections) uses TLS. Not negotiated on the
/// wire (see Docs/OFT.md §9) — both sides of a connection must be configured with the same mode, or
/// the exchange fails outright.
/// </summary>
public enum OftSecurityMode
{
    /// <summary>
    /// No TLS at all: hails are sent directly over the raw TCP connection as soon as it's formed
    /// (see Docs/OFT.md §9). No confidentiality, integrity, or authentication of either side. Only
    /// appropriate on a network already trusted by other means (see Docs/OFT.md §9).
    /// </summary>
    Trusted,

    /// <summary>
    /// TLS provides encryption (and integrity) in transit, but no authentication of either side's
    /// identity: the server presents an ephemeral, throwaway certificate it generates itself
    /// (<see cref="OftConnectionOptions.Certificate"/> is unused), and the connecting side accepts
    /// it unconditionally rather than validating it — there would be nothing meaningful to validate
    /// it against. This is the default.
    /// </summary>
    Secure,

    /// <summary>
    /// TLS provides both encryption and authentication of the server's identity: the server must
    /// present its own certificate via <see cref="OftConnectionOptions.Certificate"/>, and the
    /// connecting side validates it normally (via a caller-supplied validation callback, or the
    /// default .NET chain/hostname validation). The client is not authenticated. Not a valid mode
    /// for <see cref="IOftPeer"/>, which has no client/server delineation and so cannot express a
    /// one-sided authentication requirement (use <see cref="DualAuthentication"/> instead).
    /// </summary>
    ServerAuthentication,

    /// <summary>
    /// Mutual TLS: both sides authenticate each other. In addition to everything
    /// <see cref="ServerAuthentication"/> requires, the connecting side must present a certificate
    /// via <see cref="OftConnectionOptions.Certificate"/>, and the accepting side requests and
    /// validates it.
    /// </summary>
    DualAuthentication,
}

File: Ports/Java/src/main/java/org/blueheighliner/openframetransport/OftSecurityMode.java
package org.blueheighliner.openframetransport;

/**
 * The security mode a connection is established under (see README.md &sect;9). Mirrors the C#
 * reference implementation's {@code OftSecurityMode} enum.
 */
public enum OftSecurityMode {
    /**
     * No TLS at all: hails are sent directly over the raw TCP connection as soon as it's formed.
     * No confidentiality, integrity, or authentication of either side. Only appropriate on a
     * network already trusted by other means.
     */
    TRUSTED,

    /**
     * TLS provides confidentiality and integrity but no authentication of either side. The
     * accepting side uses a certificate it generates internally rather than one supplied by the
     * caller; the connecting side accepts whatever certificate it's presented with unconditionally.
     */
    SECURE,

    /**
     * Traditional one-way TLS: the accepting side must supply a real {@code SSLContext} carrying
     * its own certificate, which the connecting side validates normally via its own
     * {@code SSLContext}'s trust manager(s). Not a valid mode for {@link OftPeer}, which has no
     * client/server delineation and so cannot express a one-sided authentication requirement (use
     * {@link #DUAL_AUTHENTICATION} instead).
     */
    SERVER_AUTHENTICATION,

    /**
     * Mutual TLS: everything {@link #SERVER_AUTHENTICATION} requires, plus the connecting side's
     * {@code SSLContext} must also carry its own certificate (via a key manager), which the
     * accepting side requests and validates.
     */
    DUAL_AUTHENTICATION,
}
